KOGITO-2286 Pipeline modifications for release process

Kevin-Mok opened the pull request:
## PR Description
This is a PR for the following issues:
- [KOGITO-2286](https://issues.redhat.com/browse/KOGITO-2286) Kogito Runtimes deploy pipeline: Add release parameter and staging repository information
- [KOGITO-2414](https://issues.redhat.com/browse/KOGITO-2414) Kogito Runtimes deploy pipeline: Add "Set mvn version" step
- [KOGITO-2415](https://issues.redhat.com/browse/KOGITO-2415) Kogito Runtimes deploy pipeline: Add "Create PR" step
- [KOGITO-2287](https://issues.redhat.com/browse/KOGITO-2287) Kogito Runtimes promote pipeline: Create Jenkinsfile for promoting runtimes
- [KOGITO-2375](https://issues.redhat.com/browse/KOGITO-2375) Kogito Runtimes promote pipeline: Handling of Merge PR
- [KOGITO-2376](https://issues.redhat.com/browse/KOGITO-2376) Kogito Runtimes promote pipeline: Handling of Git tagging
- [KOGITO-2377](https://issues.redhat.com/browse/KOGITO-2377) Kogito Runtimes promote pipeline: Set next snapshot
- [KOGITO-2373](https://issues.redhat.com/browse/KOGITO-2373) Kogito Runtimes promote pipeline: Update examples stable branch

The first 3 were all originally part of KOGITO-2286, but I requested they be split up since the KOGITO-2414 and KOGITO-2415 parts were blocked by other tasks. But since this PR hasn't been merged yet and I've finished the code for the other 2 tasks, I asked @radtriste if it was fine to submit a single PR for all of them now, and he said it was fine. For simpler merging, the promote tasks will be added to this PR as well.
## Testing
- [deploy pipeline](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/235/)
- [promote pipeline](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2375/37/)
- branches where PR's were created and merged
  - [kogito-runtimes](https://github.com/Kevin-Mok/kogito-runtimes/commits/7.14)
  - [kogito-apps](https://github.com/Kevin-Mok/kogito-apps/commits/7.14)
  - [kogito-examples](https://github.com/Kevin-Mok/kogito-examples/commits/7.14)

## KOGITO-2286
In addition to features listed in the issue, I also discussed some refactoring of the Jenkinsfile with @radtriste which can found [here](https://github.com/Kevin-Mok/kogito-runtimes/issues/1) and [here](https://github.com/Kevin-Mok/kogito-runtimes/issues/2).  I realize "not includ[ing] fixes for issues other than the main ticket" is a requirement, but I believe the refactoring was beneficial for the implementation of the task.

## Requirements
Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

Kevin-Mok commented:
@radtriste I believe we discussed submitting a single PR for both this issue and [KOGITO-2287](https://issues.redhat.com/browse/KOGITO-2287), but I thought I'd at least open this PR for review first. Should I still add KOGITO-2287 to this PR?

radtriste commented:
@Kevin-Mok nope you can do in 2 steps. that is also ok I think

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,33 +97,38 @@ pipeline {
             }
         }
-        stage('Deploy kogito-runtimes') {
+        stage('Deploy') {
```
```suggestion
        stage('Deploy artifacts') {
```

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,33 +97,38 @@ pipeline {
             }
         }
+        stage('Archive staging repository URL') {
+            input {
```
This should be done only in the case of release

Kevin-Mok replied:
Oops. I forgot to add the check here. :sweat_smile: Fixed in latest commit.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,33 +97,38 @@ pipeline {
+                }
+            }
+            steps {
+                sh "mkdir -p archive"
```
do we want to have a specific folder for the file ?

Kevin-Mok replied:
I guess it's not necessary. I was just copying [the example](https://www.jenkins.io/doc/pipeline/examples/#archive-build-output-artifacts) where they made a folder for the archived artifacts. I put the artifacts directly into `$WORKSPACE` in the latest commit.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,33 +97,38 @@ pipeline {
+            steps {
+                sh "mkdir -p archive"
+                writeFile(file: "archive/staging-url.txt", text: "${STAGING_REPO_URL}")
+                archiveArtifacts(artifacts: "archive/staging-url.txt")
```
we should have one property file containing all deploy properties (key=value), so we can add more to that file, without having many files

radtriste replied:
Just thinking that the staging-repository is needed by the main release pipeline to be transmitted to the images & operator pipeline so that `staging-url.txt` is fine in fact.
but we could still add a property file that will be used to store the PR links in https://issues.redhat.com/browse/KOGITO-2415 and so we could add the staging-repository as well in there

Kevin-Mok replied:
I'm archiving `deployment.properties` and `staging-url.txt` in the latest commit. Do we even really need the `staging-url.txt` though? Couldn't the main release/operator pipeline read the `deployment.properties` file as well and get the URL that way?

radtriste replied:
That is open.
The way I see it:
- `deployment.properties` if for discussion between deploy and promote pipeline jobs, like "private information"
- `staging-url.txt` is public information

But I think it makes no sense to duplicate the information, so you can remove the `staging-url.txt` file and let everything in the `deployment.properties`.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,33 +97,39 @@ pipeline {
+                        def STAGING_URL_FILE = "staging-url.txt"
+                        def PROPERTIES_FILE = "deployment.properties"
+                        sh "echo ${STAGING_REPO_URL} > ${STAGING_URL_FILE}"
+                        sh "printf \"STAGING_REPO_URL=%s\" \"${STAGING_REPO_URL}\" > ${PROPERTIES_FILE}"
```
could we have this in lowercase ?

radtriste replied:
like a property:  `staging-repo.url=<VALUE>`

MarianMacik commented on Jenkinsfile.deploy.new:
```diff
@@ -34,9 +35,33 @@ pipeline {
+                    checkout(githubscm.resolveRepository("kogito-examples", "kiegroup", env.BRANCH_NAME, false))
+                }
+                // Use a separate dir for persistence to not overwrite the test results
+                sh "cp -r kogito-examples{,-persistence}"
```
Can we just manually expand this to `cp -r kogito-examples kogito-examples-persistence`? I think it looks much understandable than the current form.

Kevin-Mok replied:
Sure.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -76,37 +125,80 @@ pipeline {
                 dir("kogito-runtimes") {
                     script {
-                        maven.runMavenWithSubmarineSettings('clean deploy', true)
+                        githubscm.createBranch(NEW_BRANCH)
```
We need to `forkRepo` first, else it won't work

https://github.com/kiegroup/jenkins-pipeline-shared-libraries/blob/760f283ce035e125d8d97647894b72ee5871d455/vars/githubscm.groovy#L94

`createPR` will return the PR link that we can save into the `deployProperties` as `{project}.pr.link` key

Kevin-Mok replied:
> We need to forkRepo first, else it won't work

It seems to work without `forkRepo` though. With the current flow of `createBranch -> commitChanges -> pushObject -> createPR`, I am still able to create a PR. See [this log](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/73/console) starting from the "Create PR" stage. The resulting PR links can be found in the archived [deployment.properties](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/73/artifact/deployment.properties/*view*/).

> createPR will return the PR link that we can save into the deployProperties as {project}.pr.link key

Yep. I've set it up like so:

```groovy
deployProperties["kogito-runtimes.pr.link"] = githubscm.createPR(env.COMMIT_MSG, "master")
```

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -1,15 +1,24 @@
         MAVEN_OPTS = '-Xms1024m -Xmx4g'
+        NEW_BRANCH = "${JOB_BASE_NAME}-${BUILD_NUMBER}"
+        GIT_USER = "Jenkins CI"
+        GIT_USER_EMAIL = "jenkins@rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com"
```
These are all parameters now in the latest commit.

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -1,15 +1,24 @@
     }
     environment {
         MAVEN_OPTS = '-Xms1024m -Xmx4g'
+        NEW_BRANCH = "${JOB_BASE_NAME}-${BUILD_NUMBER}"
```
`NEW_BRANCH` is now an optional parameter (as per [your other comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r447545529)) that becomes the new branch name if set. If not, the new branch name is `PROJECT_VERSION`.

The description for the parameter mentions it's only used if `RELEASE` since a branch is only needed if it's a release and PR's are being made. `RELEASE` also now guarantees `PROJECT_VERSION` is specified as per [your other comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r447559763).

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -1,15 +1,24 @@
+        NEW_BRANCH = "${JOB_BASE_NAME}-${BUILD_NUMBER}"
+        GIT_USER = "Jenkins CI"
+        GIT_USER_EMAIL = "jenkins@rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com"
+        COMMIT_MSG = "Update Maven version for release"
```
I went with `COMMIT_MSG = "Update project version to ${params.PROJECT_VERSION}"`.

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -76,37 +125,80 @@ pipeline {
         always {
             junit '**/target/surefire-reports/**/*.xml, **/target/failsafe-reports/**/*.xml'
+            script {
+                if (params.RELEASE) {
```
Done. This makes more sense now that `deployment.properties` stores the  'RELEASE' and  'PROJECT_VERSION' parameters and will never be empty.

Kevin-Mok commented:
@radtriste I've added the requested changes [in an extra commit](https://github.com/kiegroup/kogito-runtimes/pull/570/commits/fc9739cc6c288dfdbb9f4cae894cda21a0c97ab3) that I can squash later before the PR, so the diff is easier to see. 

### Testing
There's a bunch of branching behavior now between `RELEASE`/not `RELEASE` and empty `PROJECT_VERSION`/`NEW_BRANCH_NAME` or not. I spent a bit of time trying to setup [jenkins-spock](https://github.com/ExpediaGroup/jenkins-spock) to test this Jenkinsfile, as @Ginxo had me do for jenkins-pipeline-shared-libraries, but it turned out to be more complicated to setup/use than I thought it would be... :sweat_smile:

I ended up manually testing all the branching cases, but perhaps it would be worth considering implementing in the future?

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +165,85 @@ pipeline {
                     script {
-                        maven.runMavenWithSubmarineSettings('clean deploy', true)
+                        githubscm.createBranch(env.NEW_BRANCH)
+                        githubscm.commitChanges(params.GIT_USER, params.GIT_USER_EMAIL, env.COMMIT_MSG, "pom.xml")
```
we have multiple pom which are updated by the `updateMavenVersion` so they should also be added

Kevin-Mok replied:
Would it just be `"pom.xml */pom.xml"`?

radtriste replied:
`**/pom.xml` should be or ?

Kevin-Mok replied:
So, glob won't work with functions. If you pass `'**/pom.xml'` to `commitChanges()` as the `filesToAdd` parameter, it turns `'**/pom.xml'` into a string literal such that the shell command becomes `git add '**/pom.xml'` and won't match the files. See [this build log](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/118/execution/node/138/log/) which was ran with `commitChanges(_, _, _, '**/pom.xml')`.

If you run it with the default parameter of `'--all'` though, it doesn't get turned into a string literal and [runs `git add --all`](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/117/execution/node/138/log/). I'm not sure how to get around this problem. :thinking: @Kaitou786?

Kevin-Mok replied:
So after some more testing, it seems that the problem is not that `**` doesn't work in quotes because apparently it does; it's just that it `**` behaves strangely in Jenkins.  With [the files `a/b/c d/e/c f/c c`](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/158/console), `git add **/c` becomes `git add f/c` only matches `f/c`.  With [the files `a/b/c d/e/c c`](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/159/console), `git add **/c` becomes `git add '**/c'` matches `a/b/c` and `d/e/c`. Basically, it seems to be inconsistent in its behavior and unreliable. So, what I came up with is to use [`find` with command substitution](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/155/console) which seems to get the job done.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +165,85 @@ pipeline {
+            steps {
+                script {
+                    emailext body: "${env.PIPELINE_NAME} #${env.BUILD_NUMBER} needs a staging repository URL: ${env.BUILD_URL}input",
+                             subject: "${env.PIPELINE_NAME} #${env.BRANCH_NAME}",
```
```suggestion
                             subject: "${env.PIPELINE_NAME} #${env.BUILD_BRANCH_NAME}",
```

Kevin-Mok replied:
Shouldn't it actually be `env.BRANCH_NAME` since we're checking out that branch and building it?

https://github.com/kiegroup/kogito-runtimes/blob/1240c1fe000b3326a949ce159b821d9c1ab0d532/Jenkinsfile.deploy.new#L39

radtriste replied:
yeah because we set 
```
if (env.BRANCH_NAME != "") {
                        // Switch to branch if not on a multibranch pipeline
                        env.BRANCH_NAME = params.BUILD_BRANCH_NAME
                    }
```
let it with `BRANCH_NAME`, I will surely PR later to unify this

Kevin-Mok replied:
I used `getBuildBranch()`.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +165,85 @@ pipeline {
     post {
         always {
+            script {
+                def propertiesStr = ""
```
could we do something like this ?

```
propertiesStr = deployProperties.collect{ entry ->  "${entry.key}=${entry.value}" }.join("\n")
```

Kevin-Mok replied:
Fancy. :smile:

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -31,12 +42,74 @@ pipeline {
                     }
+                    deployProperties['release'] = params.RELEASE
+                    deployProperties['project-version'] = params.PROJECT_VERSION
+                    env.NEW_BRANCH = params.NEW_BRANCH_NAME
```
> This is only used in Create PR stage, so I think we can move this part to that stage.

There will be an empty `deployment.properties` if the `assert` fails. I guess that's fine?

> Also do we need the parameter NEW_BRANCH_NAME as it will be done only in the case of a RELEASE, so the version should be set?

I was under the impression that [your other comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r447545529) suggested making `NEW_BRANCH` a parameter. :thinking:

radtriste replied:
> > This is only used in Create PR stage, so I think we can move this part to that stage.
> 
> There will be an empty `deployment.properties` if the `assert` fails. I guess that's fine?
> 
I was referencing the `NEW_BRANCH` part that should be moved to `Create PR` stage.
The set of deploy properties can stay here

> > Also do we need the parameter NEW_BRANCH_NAME as it will be done only in the case of a RELEASE, so the version should be set?
> 
> I was under the impression that [your other comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r447545529) suggested making `NEW_BRANCH` a parameter.
Oh I was mentioning the fact that `GITHUB_USER` and `GITHUB_USER_EMAIL` could be environment variables configured globally in Jenkins and not specifically here. Sorry if I was not clear ... no need to have them as params.  `GITHUB_USER` and `GITHUB_USER_EMAIL` can stay as env ... Sorry for the extra work there :S

Kevin-Mok replied:
I've reverted the branch name to just the `PROJECT_VERSION` parameter and removed the `GITHUB_USER` and `GITHUB_USER_EMAIL` parameters in favor of the globally set ones.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -31,12 +42,74 @@ pipeline {
                         env.BRANCH_NAME = params.BUILD_BRANCH_NAME
                     }
+                    deployProperties['release'] = params.RELEASE
+                    deployProperties['project-version'] = params.PROJECT_VERSION
```
can we set the project version only if not empty ?

Kevin-Mok commented:
@radtriste I've added the requested changes [in another commit](https://github.com/kiegroup/kogito-runtimes/pull/570/commits/6e68c003b19c3d0097ba65a2dccc142a7e8c478f) for diff purposes. Only having issues with the globbing as mentioned in [my response to the review](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r449268738).

## Bot Account for PR's
I'll need to get the credentials ID for the bot account we're going to use. @Kaitou786 sent me a bot account that could be used, but I'm not sure who has access to it that can create the access token. I also looked through the credentials in Kogito folder and didn't seem to see a bot-looking account on there.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -1,15 +1,26 @@
+        PIPELINE_NAME = "Kogito Runtimes deploy pipeline"
         MAVEN_OPTS = '-Xms1024m -Xmx4g'
+        PROPERTIES_FILE = "deployment.properties"
+        FILES_TO_ADD = '**/pom.xml'
```
not sure that one is really needed as en var. we could just have it as local var in stage `Create PR`

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +163,85 @@ pipeline {
+            }
+            steps {
+                script {
+                    emailext body: "${env.PIPELINE_NAME} #${env.BUILD_NUMBER} needs a staging repository URL: ${env.BUILD_URL}input",
```
Set pipeline name as a local var as only used here

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +163,85 @@ pipeline {
+            steps {
+                script {
+                    emailext body: "${env.PIPELINE_NAME} #${env.BUILD_NUMBER} needs a staging repository URL: ${env.BUILD_URL}input",
+                             subject: "${env.PIPELINE_NAME} #${env.BUILD_BRANCH_NAME}",
```
I would add at the end `RELEASE` in the subject so we know it is in release process

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +163,85 @@ pipeline {
-                        maven.runMavenWithSubmarineSettings('clean deploy', true)
+                        githubscm.forkRepo(env.BOT_CREDENTIALS_ID)
+                        githubscm.createBranch(params.PROJECT_VERSION)
+                        githubscm.commitChanges(env.GIT_COMMITTER_NAME, env.GIT_COMMITTER_EMAIL, env.COMMIT_MSG, env.FILES_TO_ADD)
```
`COMMIT_MSG` could be a local var as well IMO

Kevin-Mok replied:
I believe I tried this initially but since each of these scripts are separated into their own `dir` block, they can't access a shared local variable. Also, I will refactor those very similar scripts into a function.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -31,12 +42,72 @@ pipeline {
                         // Switch to branch if not on a multibranch pipeline
                         env.BRANCH_NAME = params.BUILD_BRANCH_NAME
                     }
+                    if (params.RELEASE) {
```
maybe add also the `BUILD_BRANCH_NAME` as `git.branch` property ?

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -31,12 +42,72 @@ pipeline {
+        }
+        stage('Update project version'){
+            when {
+                expression { return params.PROJECT_VERSION != '' }
```
IMO, it'd be unnecessary for this deploy Jenkinsfile because the project version and release would always be parameters? We were discussing how they may be retrieved differently in the promote pipeline, and helper functions make more sense to me there. However, since it's simply going to always going to return the parameter itself in this case, it seems more readable and simplified to me if the direct parameter was used here?

radtriste replied:
Just that I find it more readable and we unify both deploy and promote in their way of writing the code.

Kevin-Mok commented:
@radtriste Latest fixes can be found in [this commit](https://github.com/kiegroup/kogito-runtimes/pull/570/commits/58f3b233f3a99c978cc8833bd0f4454a29437d74). 

### Changes
- use reworked shared library functions to set project version
- refactored code block that creates PR to function
- got rid of all the environment variables and moved them to their local usage blocks

Ginxo commented:
@Kevin-Mok remember you can always move your Jenkinsfile in https://github.com/kiegroup/jenkins-pipeline-shared-libraries/tree/master/test/resources (as Jenkinsfile.kogito-runtimes for instance) and create a test for it (KogitoRuntimesPipelineSpec loading the Jenkinsfile instead of a groovy script). (just suggesting)

radtriste commented:
@Ginxo So you are suggesting to test by checking out the shared libs, move the pipeline in resources and a test and execute the whole test ?

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -1,12 +1,16 @@
 @Library('jenkins-pipeline-shared-libraries')_
 
+def deployProperties = [:]
```
as discussed in https://github.com/kiegroup/kogito-cloud-operator/pull/420/files#r450517728, it should be

```suggestion
deployProperties = [:]
```

to have global scope

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -31,12 +35,75 @@ pipeline {
+                    if (params.RELEASE) {
+                        assert params.PROJECT_VERSION != ''
+                    }
+                    if (params.BUILD_BRANCH_NAME != '') {
```
`BUILD_BRANCH_NAME` will never be empty (default is master)

Kevin-Mok replied:
I copied your line: `setDeployPropertyIfNeeded('git.branch', getBuildBranch())`.

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -96,41 +159,92 @@ pipeline {
         always {
+            script {
+                def propertiesStr = deployProperties.collect{ entry ->  "${entry.key}=${entry.value}" }.join("\n")
+                def propertiesFile = "deployment.properties"
+                writeFile(text: propertiesStr, file: propertiesFile)
```
```suggestion
                writeFile(text: propertiesStr, file: "deployment.properties")
```

Ginxo commented:
_So you are suggesting to test by checking out the shared libs, move the pipeline in resources and a test and execute the whole test ?_ -> @radtriste I'm suggesting you have the chance to do it in case you want to cover your pipeline ;)

Kevin-Mok commented:
@radtriste Latest fixes can be found in [this commit](https://github.com/kiegroup/kogito-runtimes/pull/570/commits/b55f13c98ef999123693d7f213a4acd2da00b07b). 

### Changes
- added getter/setter functions from @radtriste's PR
- moved forking/creating branch to right after cloning (if release)
- fixed issue of `find` adding `pom.xml`'s in `.gitignore` (see [this build](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/191/execution/node/309/log/)) with `addNotIgnoredPoms()`
  - based on [this SO answer](https://stackoverflow.com/a/59888964/8811872) 
    - [alternative solutions](https://unix.stackexchange.com/q/358270/260572) exist but I felt the above answer was the most readable
    - modified answer based on [ShellCheck suggestions](https://github.com/koalaman/shellcheck/wiki/SC2044#for-posix)

### Testing
I ran an initial shorter test of only working with kogito-runtimes, and the first build had [22 failed tests](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/199/#showFailuresLink). Upon running it again though, there [were no failures](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/203). 

With the full build I ran, it failed due to there being [no changed files in the commit](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/211/execution/node/336/log/), even after updating the project version and building. So, does no PR get created for kogito-apps then, @radtriste?

radtriste commented:
@Kevin-Mok 
For the update of the version, it seems like once runtimes version is updated, you will need to install in local maven `mvn -B clean install -DskipTests` and then you should be able to change the maven version of kogito-apps.
This was the fix done in shell script: https://github.com/kiegroup/kogito-runtimes/blob/feffb9bab334d2e151ef6aca44fc4d44b29f120d/scripts/release/02-update-version-all.sh#L92

btw, there is also a problem updating the parent version, the command should be:
`mvn -B -N -e versions:update-parent -Dfull '-DparentVersion=[0.12.0]' -DallowSnapshots=true -DgenerateBackupPoms=false`
instead of
`mvn -B -N -e versions:update-parent -Dfull -DparentVersion=0.12.0 -DallowSnapshots=true -DgenerateBackupPoms=false`
or we should have a possibility to set the `-DallowSnapshots` to false as parameter of the shared lib method ...
Can you make an update in shared libs ?

Kevin-Mok commented:
@radtriste I've squashed/rebased my commits and created a separate branch for diffs. 

## Jenkinsfile.deploy
Latest fixes for Jenkinsfile.deploy can be found in [this commit](https://github.com/Kevin-Mok/kogito-runtimes/commit/58da40aab666e15c4bd153723359392ead3ee6ff). 

### Changes
- able to update `kogito-apps`' version using `mvn -B clean install -DskipTests` and updated Maven functions from [KOGITO-2708](https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/74)
- use updated `createPR` function from [KOGITO-2711](url) to add build tag/URL to PR body

### Testing
I was able to get a [full passing build](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/225/) with the `Jenkinsfile.deploy` which produced the following PR's:
- [kogito-runtimes](https://github.com/Kevin-Mok/kogito-runtimes/pull/8)
- [kogito-apps](https://github.com/Kevin-Mok/kogito-apps/pull/2)
- [kogito-examples](https://github.com/Kevin-Mok/kogito-examples/pull/1)

### Questions
1. I noticed that the `sed` command found in `02-update-version-all.sh` looks for `<version.org.kie.kogito>`, which doesn't occur in `pom.xml`. Is it actually needed?
https://github.com/kiegroup/kogito-runtimes/blob/feffb9bab334d2e151ef6aca44fc4d44b29f120d/scripts/release/02-update-version-all.sh#L90
2. I ran my full build with all the default `allowSnapshots` parameters to the update Maven functions set to `false`. Which actually need to be `true`?

## [[KOGITO-2287]](https://issues.redhat.com/browse/KOGITO-2287) Create Jenkinsfile for promoting runtimes
Task can be found in [Jenkinsfile.promote](https://github.com/Kevin-Mok/kogito-runtimes/blob/KOGITO-2286-diff/Jenkinsfile.promote).  Borrows heavily from Jenkinsfile.promote in [KOGITO-2290](https://github.com/kiegroup/kogito-cloud-operator/pull/420).  Also left some review comments for things I found while going through the file.

### Testing
[Example build log](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2287/8/console).  [Zulip message](https://kie.zulipchat.com/#narrow/stream/236603-kogito-ci/topic/Test.20%28kmok%29.20-.20Kogito.20Runtimes.20promote.20pipeline/near/203345351) from build.

radtriste commented:
1. Indeed, seems it is not needed anymore. Good catch.
2. They can be let as false for the release pipeline as we are putting a release version and not snapshot

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,136 @@
+                }
+            }
+        }
+        stage('Approve promotion') {
```
This stage is done to let the pipeline knows that the staging repository has been promoted to Maven Central.
I would set the name of the stage with a question mark, like `Is staging repository promoted ?`

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,136 @@
+                    withCredentials([string(credentialsId: "KOGITO_CI_EMAIL_TO", variable: 'ZULIP_EMAIL')]) {
+                        emailext body: "${pipelineName} #${env.BUILD_NUMBER} is ready for promotion.\n" +
+                                 "The staging repository can be found at: ${getDeployProperty('staging-repo.url')}\n" +
+                                 "Approve the promotion here: ${env.BUILD_URL}input",
```
```suggestion
                                 "Please confirm promote this repository and then confirm the promotion here: ${env.BUILD_URL}input",
```

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,136 @@
+
+void readDeployProperties(){
+    if (params.DEPLOY_BUILD_URL != null){
+        withCredentials([usernamePassword(credentialsId: 'rhba-jenkins-token', usernameVariable: 'JENKINS_USER', passwordVariable: 'JENKINS_TOKEN')]) {
```
As stated here: https://github.com/kiegroup/kogito-cloud-operator/pull/420/files?file-filters%5B%5D=.promote#r452030869
I am not sure we need the TOKEN

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -29,14 +33,90 @@ pipeline {
+                dir("kogito-runtimes") {
+                    script {
+                        maven.mvnVersionsSet(getProjectVersion())
+                        sh "mvn -B clean install -DskipTests"
```
```suggestion
                        // Step needed to have runtimes artifacts into the local repository, in order to set apps version correctly
                        sh "mvn -B clean install -DskipTests"
```

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+        stage('Merge deploy PR') {
+            when {
+                expression {
+                    return isRelease() && 
+                           getDeployProperty('kogito-runtimes.pr.link') != '' && 
+                           getDeployProperty('kogito-apps.pr.link') != '' && 
+                           getDeployProperty('kogito-examples.pr.link') != ''
```
I think we should have 3 different stages here, one for each project for the `Merge PR`

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+
+void tagLatest(String repo) {
+    githubscm.tagRepository(getGitTag(), BUILD_TAG)
+    githubscm.pushObject("origin", BUILD_TAG)
```
push of a tag should be `git push --tags`

and we should push the `getGitTag()`

radtriste replied:
btw, we would need to have latest change before tagging or ?

Kevin-Mok replied:
> btw, we would need to have latest change before tagging or ?

I don't think so because in the previous "Merge deploy PR" stage, we check out the repo, merge the PR and push to `origin`, so the local version should already be the most updated.

radtriste replied:
so when merging the PR, changes are automatically merged also on local ?

radtriste replied:
ok just saw the other comment ...

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+    githubscm.pushObject("origin", getBuildBranch())
+}
+
+void tagLatest(String repo) {
```
why `tagLatest` ?

Kevin-Mok replied:
I was basing it on your `TODO` comment in the "Git tag" stage: `// TODO Retrieve latest BUILD_BRANCH_NAME / GIT_AUTHOR ?`

radtriste replied:
`tagRemoteRepo` ?

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+            }
+            steps {
+                dir("kogito-runtimes-bot") {
+                    script {
+                        commitAndCreatePR("kogito-runtimes")
+                    }
+                }
+                dir("kogito-apps-bot") {
+                    script {
+                        commitAndCreatePR("kogito-apps")
+                    }
+                }
+                dir("kogito-examples") {
+                    script {
+                        commitAndCreatePR("kogito-examples")
+                    }
+                }
+            }
+        }
+        stage('Merge snapshot PR') {
+            when {
+                expression {
+                    return isRelease() && 
+                           getDeployProperty('kogito-runtimes.pr.link') != '' && 
+                           getDeployProperty('kogito-apps.pr.link') != '' && 
+                           getDeployProperty('kogito-examples.pr.link') != ''
+                }
+            }
+            steps{
+                dir("kogito-runtimes") {
+                    mergeAndPush("kogito-runtimes")
+                }
+                dir("kogito-apps") {
+                    mergeAndPush("kogito-apps")
+                }
+                dir("kogito-examples") {
+                    mergeAndPush("kogito-examples")
+                }
+            }
+        }
```
I think we could have one stage by project doing those steps instead of having many staging for each action
It would be called something like `Set and push next snapshot` and include so:

- Prepare
- Set next snapshot
- Create & Merge PR

Kevin-Mok replied:
This is a much cleaner approach.

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+
+void checkoutRepo(String repo) {
+    checkout(githubscm.resolveRepository(repo, "kiegroup", getBuildBranch(), false))
+    sh "git checkout ${getBuildBranch()}"
```
why do you checkout the `getBuildBranch()` after ? isn't it done by the first checkout ?

Kevin-Mok replied:
It's on a detached branch after checkout. See [this build log](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/test/16/console) where the following script was run (end of line comments are output):
```
checkout(githubscm.resolveRepository("jenkins-test", "Kevin-Mok", "master", false))
sh """
git branch          # (detached from c73ac4d)
git checkout master # Switched to a new branch 'master'
git branch          # master
"""
```

radtriste replied:
oki. Can you add a comment so we understand that quickly ? :)

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+
+void mergeAndPush(String repo) {
+    githubscm.mergePR(getDeployProperty("${repo}.pr.link"))
+    githubscm.pushObject("origin", getBuildBranch())
```
why do we need to push after the merging of PR ?

Kevin-Mok replied:
`hub merge` only creates a local commit. From [its man page](https://hub.github.com/hub-merge.1.html):

> [`hub merge`] creates a local merge commit in the current branch, but does not actually change the state of the pull request. However, the pull request will get auto-closed and marked as "merged" as soon as the newly created merge commit is pushed to the default branch of the remote repository.

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+                        assert getProjectVersion() != ''
+                    }
+                    if (getGitTag() != ""){
+                        assert getBuildBranch() != ''
```
build branch should not be empty anyway here or ?

Kevin-Mok replied:
I just copied an older version of your "Initialization" stage. But, I see that it's gone now, so I will remove it too. :smile:

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,282 @@
+            when {
+                expression {
+                    return isRelease() && 
+                           getDeployProperty('kogito-runtimes.pr.link') != '' && 
+                           getDeployProperty('kogito-apps.pr.link') != '' && 
+                           getDeployProperty('kogito-examples.pr.link') != ''
```
is that needed here ?

Kevin-Mok replied:
This will be gone when I create separate stages for each project as you suggested in [your other comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r452760513).

Kevin-Mok commented:
@radtriste I've added the code for KOGITO-{2375,2376,2377} into this PR such that this PR now contains all the tasks related to the deploy/promote pipeline for runtimes. 

## Pipeline Tests
I ran full test builds for both the deploy and promote pipelines. The full build for the deploy pipeline can be [found here](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2415/235/).  Then, a `deployment.properties` from a build was passed into a [promote pipeline build](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2375/37/console).  The build actually ends up failing when trying to create the snapshot PR for apps. This is because the PR from the deploy pipeline is merged such that it's trying to find an artifact that doesn't exist in the Maven repositories since this is only a test build:
```
[INFO] Downloading from central: https://repo.maven.apache.org/maven2/org/kie/kogito/kogito-runtimes/7.14.1/kogito-runtimes-7.14.1.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.kie.kogito:kogito-apps:7.14.1: Could not find artifact org.kie.kogito:kogito-runtimes:pom:7.14.1 in central (https://repo.maven.apache.org/maven2) and 'parent.relativePath' points at wrong local POM @ line 7, column 11
```

During an actual release build, I guess the runtimes pom will be pushed to the Maven repository, and this won't be an issue? Other than this issue, everything seems to work as expected.

## Testing Helper Functions with jenkins-spock
I wanted to use jenkins-spock to test some helper functions and tried setting it up as an additional module here.  I basically copied over the `pom.xml` from the shared libs into a `jenkins-spock-tests` directory, added it as a module and copied the directory structure of the shared libs. It works fine if I run `mvn test` in the `jenkins-spock-tests`, but it doesn't run the tests if I run `mvn test` in the base project directory. 

I noticed that there's a difference in the console output.  When running `mvn test` from the `jenkins-spock-tests/` directory, its able to find the file to compile:
```
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ jenkins-spock-tests ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/kevin/coding/kogito/kogito-runtimes/jenkins-spock-tests/target/test-classes
[INFO] 
[INFO] --- gmavenplus-plugin:1.6.1:compileTests (groovy) @ jenkins-spock-tests ---
[INFO] Using Groovy 2.4.11 to perform compileTests.
[WARNING] Configuration script file (/home/kevin/coding/kogito/kogito-runtimes/jenkins-spock-tests/config.groovy) doesn't exist.  Ignoring configScript parameter.
[INFO] Compiled 1 file.
```

Where as running `mvn test` in the base directory yields:
```
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ jenkins-spock-tests ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- gmavenplus-plugin:1.6.1:compileTests (groovy) @ jenkins-spock-tests ---
[INFO] No sources specified for compilation.  Skipping.
```

I tried adding this line to the `pom.xml`: `<tests.surefire.include.groovy>**/*Test.groovy</tests.surefire.include.groovy>`.  But, that didn't help. I also tried adding [this configuration section](https://github.com/Kevin-Mok/kogito-runtimes/blob/0cd01339cf2487143897f2d27d71e32e02ebafe2/pom.xml#L1332-L1361) for the `gmavenplus-plugin` from the shared lib pom, but that didn't help either. @Ginxo, do you have any idea how this can be fixed?

I'm also getting this error since I'm using a symbolic link to link the Jenkinsfile, but it doesn't seem to affect the test results:
```
21:25:14.516 [main] DEBUG org.reflections.Reflections - could not scan file Jenkinsfile.promote in url file:/home/kevin/coding/kogito/kogito-runtimes/jenkins-spock-tests/target/test-classes/ with scanner SubTypesScanner
org.reflections.ReflectionsException: could not create class object from file Jenkinsfile.promote
	at org.reflections.scanners.AbstractScanner.scan(AbstractScanner.java:32)
	at org.reflections.Reflections.scan(Reflections.java:253)
	at org.reflections.Reflections.scan(Reflections.java:202)
	at org.reflections.Reflections.<init>(Reflections.java:123)
	at com.homeaway.devtools.jenkins.testing.LocalProjectPipelineExtensionDetector.getClassesOfTypeInPackage(LocalProjectPipelineExtensionDetector.java:61)
	at com.homeaway.devtools.jenkins.testing.LocalProjectPipelineExtensionDetector$getClassesOfTypeInPackage.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at com.homeaway.devtools.jenkins.testing.JenkinsPipelineSpecification.setupSpec(JenkinsPipelineSpecification.groovy:891)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:188)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481) at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.doRunSetupSpec(BaseSpecRunner.java:180)
	at org.spockframework.runtime.BaseSpecRunner$3.invoke(BaseSpecRunner.java:164)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.runSetupSpec(BaseSpecRunner.java:157)
	at org.spockframework.runtime.BaseSpecRunner.doRunSetupSpec(BaseSpecRunner.java:177)
	at org.spockframework.runtime.BaseSpecRunner$3.invoke(BaseSpecRunner.java:164)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.runSetupSpec(BaseSpecRunner.java:157)
	at org.spockframework.runtime.BaseSpecRunner.runSetupSpec(BaseSpecRunner.java:152)
	at org.spockframework.runtime.BaseSpecRunner.doRunSpec(BaseSpecRunner.java:97)
	at org.spockframework.runtime.BaseSpecRunner$1.invoke(BaseSpecRunner.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.runSpec(BaseSpecRunner.java:76)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:67)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: /home/kevin/coding/kogito/kogito-runtimes/jenkins-spock-tests/target/test-classes/Jenkinsfile.promote (No such file or directory)
	at org.reflections.vfs.SystemFile.openInputStream(SystemFile.java:34)
	at org.reflections.adapters.JavassistAdapter.getOfCreateClassObject(JavassistAdapter.java:98)
	at org.reflections.adapters.JavassistAdapter.getOfCreateClassObject(JavassistAdapter.java:24)
	at org.reflections.scanners.AbstractScanner.scan(AbstractScanner.java:30)
	... 43 common frames omitted
Caused by: java.io.FileNotFoundException: /home/kevin/coding/kogito/kogito-runtimes/jenkins-spock-tests/target/test-classes/Jenkinsfile.promote (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at org.reflections.vfs.SystemFile.openInputStream(SystemFile.java:32)
	... 46 common frames omitted
```

radtriste commented:
> During an actual release build, I guess the runtimes pom will be pushed to the Maven repository, and this won't be an issue? Other than this issue, everything seems to work as expected. Here are the test branches that were used where the deploy and snapshot PR's were merged:

Yes, artifacts 7.14.1 should be in Maven Central already as we ask for input when artifacts are promoted :)

> Testing Helper Functions with jenkins-spock

Nice ! If I have time today I will give a look to check what is the problem ...

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,272 @@
+                }
+            }
+        }
+        stage('Create/merge examples snapshot PR'){
```
```suggestion
        stage('Set examples next snapshot version'){
```

radtriste commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,272 @@
+        }
+        stage('Create/merge examples snapshot PR'){
+            when {
+                expression { return isRelease() && getProjectVersion() != "" }
```
```suggestion
                expression { return isRelease() }
```

radtriste commented:
@Kevin-Mok I think you are missing the `${project.basedir}` in the different paths.
Try:
```
<build>
        <pluginManagement>
            <plugins>
                <plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-surefire-plugin</artifactId>
                    <version>${surefire.pluginVersion}</version>
                    <executions>
                        <execution>
                            <id>default-test</id>
                            <goals>
                                <goal>test</goal>
                            </goals>
                            <configuration>
                                <forkCount>1</forkCount>
                                <includes>
                                    <include>${test.pattern}</include>
                                </includes>
                                <useManifestOnlyJar>false</useManifestOnlyJar>
                                <systemPropertyVariables>
                                    <root.loglevel>${test.loglevel}</root.loglevel>
                                    <root.appender>Stdout</root.appender>
                                    <test.loglevel>${test.loglevel}</test.loglevel>
                                    <logdir>${logdir}</logdir>
                                </systemPropertyVariables>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>
                <plugin>
                    <groupId>org.codehaus.gmavenplus</groupId>
                    <artifactId>gmavenplus-plugin</artifactId>
                    <version>${groovy.gmaven.pluginVersion}</version>
                    <executions>
                        <execution>
                            <id>groovy</id>
                            <goals>
                                <goal>addSources</goal>
                                <goal>addTestSources</goal>
                                <goal>generateStubs</goal>
                                <goal>generateTestStubs</goal>
                                <goal>compile</goal>
                                <goal>compileTests</goal>
                                <goal>removeStubs</goal>
                                <goal>removeTestStubs</goal>
                            </goals>
                            <configuration>
                                <configScript>${project.basedir}/config.groovy</configScript>
                                <sources>
                                    <source>
                                        <directory>${project.basedir}/src/test/resources</directory>
                                        <includes>
                                            <include>Jenkinsfile.*</include>
                                        </includes>
                                    </source>
                                </sources>
                                <testSources>
                                    <testSource>
                                        <directory>${project.basedir}/src/test/vars</directory>
                                        <includes>
                                            <include>**/*.groovy</include>
                                        </includes>
                                    </testSource>
                                </testSources>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>
            </plugins>
        </pluginManagement>
        <plugins>
            <plugin>
                <groupId>org.codehaus.gmavenplus</groupId>
                <artifactId>gmavenplus-plugin</artifactId>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
            </plugin>
        </plugins>
        <sourceDirectory>${project.basedir}/src/test/resources</sourceDirectory>
        <testResources>
            <testResource>
                <directory>${project.basedir}/src/test/resources</directory>
                <includes>
                    <include>Jenkinsfile.*</include>
                </includes>
            </testResource>
            <testResource>
                <directory>${project.basedir}/src/test/vars</directory>
                <includes>
                    <include>**/*.groovy</include>
                </includes>
            </testResource>
        </testResources>
    </build>
```
worked for me

Also I get the same error but indeed, it does not seem to affect the test results ...

radtriste commented on jenkins-spock-tests/src/test/resources/Jenkinsfile.promote:
```diff
@@ -0,0 +1 @@
+../../../../Jenkinsfile.promote
```
what about copying the `Jenkinsfile.promote` to `resources` before compiling and testing, via `copy-resources` plugin ?

Kevin-Mok commented on Jenkinsfile.deploy.new:
```diff
@@ -111,41 +186,125 @@ pipeline {
+                    def pipelineName = "Kogito Runtimes deploy pipeline"
+                    withCredentials([string(credentialsId: "KOGITO_CI_EMAIL_TO", variable: 'ZULIP_EMAIL')]) {
+                        emailext body: "${pipelineName} #${env.BUILD_NUMBER} is ready for release and needs the staging repository URL. Please provide it at: ${env.BUILD_URL}input",
+                                 subject: "${pipelineName} #${getBuildBranch()}",
```
Should it be the build branch after it or the project version then?

radtriste replied:
good question.
I would keep one stream per release branch. so the build branch should be enough for now. We can still easily update this later

Kevin-Mok commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,272 @@
+            when {
+                expression {
+                    return isRelease() &&
+                           getDeployProperty('kogito-runtimes.pr.link') != '' &&
+                           getGitTag() != ''
```
So for here and the other stages, only check `isRelease()` and that's it?

radtriste replied:
yep, and add `if` clause for each of the steps in the  script that will execute if the value is not empty

Kevin-Mok commented:
@radtriste I've made the changes requested and got the jenkins-spock module working. The diff can be found [here](https://github.com/Kevin-Mok/kogito-runtimes/commit/340619e84d9a4b8ad0d172326b6c2b32244e80ea).

I left some responses regarding some changes I'm not completely sure about. The module problem was indeed solved by adding `${project.basedir}`.  I think I used that to add the Jenkinsfile from the base directory as a resource. I'm not completely sure since for whatever reason, it doesn't work if I remove the symbolic link, but I don't get the error any more... :man_shrugging:

Kevin-Mok commented:
@radtriste I made minor logic changes as mentioned in [this comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r455666214); I just moved the `if` checks to the helper functions. Diff can be found [here](https://github.com/Kevin-Mok/kogito-runtimes/commit/e727b4ba9396421350b6a00fe178043eebb4e240).

Kevin-Mok commented:
@radtriste I added [KOGITO-2373](https://issues.redhat.com/browse/KOGITO-2373) to this PR, which I've highlighted in [this diff](https://github.com/Kevin-Mok/kogito-runtimes/commit/3e6b33b062b447cb5aabce06c94249200c59559c#diff-55fc28b547fad1a92acc0c1a451670a5R92-R96).  The rest of the changes are the minor logic changes as mentioned in [this comment](https://github.com/kiegroup/kogito-runtimes/pull/570#discussion_r455666214); I just moved the `if` checks to the helper functions.

## KOGITO-2373 Test
I force-pushed a test branch ([7.14.3-SNAPSHOT](https://github.com/Kevin-Mok-Bot/kogito-examples/tree/7.14.3-SNAPSHOT)) onto [`stable`](https://github.com/Kevin-Mok-Bot/kogito-examples/tree/stable) in my forked kogito-examples on my account using the highlighted core code of KOGITO-2373. You can find the build log [here](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2373/1/console).

radtriste commented:
all good for me. will have a run tomorrow in the release pipeline global test

Kevin-Mok commented:
@radtriste, I asked @cristianonicolai about force pushing the `stable` branch on kogito-examples for [KOGITO-2373](https://issues.redhat.com/browse/KOGITO-2373), and he said that force pushing is disabled for that branch.

@evacchi, do you have any alternate ideas for how we can automate updating the `stable` branch from within a Jenkins build?  This is how it's [currently being done](https://github.com/Kevin-Mok/kogito-runtimes/blob/cb49119d71ab556b6c92011c059db5da8cdd44f9/Jenkinsfile.promote#L92-L96).

radtriste commented:
@cristianonicolai @evacchi How is done the update of the stable branch for now ? by hand on GH UI ?

Kevin-Mok commented:
@radtriste From Cristiano:

> Edoardo has been doing this manually, I believe he ends up recreating the branch via admin page from github itself

Kevin-Mok commented:
@radtriste, I refactored Jenkinsfile.deploy and implemented a workaround not being able to force push branches.

## Refactoring Jenkinsfile.deploy
I didn't realize that the `dir` syntax could be used in helper functions, so I refactored a bunch of the stages that had `dir` for each of the repos into the helper functions like you did in your PR for Jenkinsfile.deploy. I also copied your `mavenInstall`/`mavenDeploy` helper functions. This refactoring diff can be [found here](https://github.com/Kevin-Mok/kogito-runtimes/commit/1c5b85a979267cc2ebaac55182769eb3e10e6479).

I didn't do this for Jenkinsfile.promote because I believe it's more readable in its current state with the `dir` blocks showing which steps are being run in the bot directory vs. main one.

## Workaround for Force Pushing
I found out that [force pushing to kiegroup repositories](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2373/3/execution/node/27/log/) is not allowed on any branch, but [deleting branches](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/kmok/job/KOGITO-2373/4/execution/node/27/log/) are. So, I basically recreated the steps that @evacchi currently does to update `stable`:
1. Switch default branch from `stable` to a temporary branch (i.e. `master`).
2. Delete `stable` branch.
3. Push latest commit from release branch to new `stable` branch.
4. Switch `stable` back to be the default branch.

The code for this workaround can be found in [this commit](https://github.com/Kevin-Mok/kogito-runtimes/commit/fc292400c45cd3067eeb6190d6888d525194a9f3).

radtriste commented on Jenkinsfile.deploy.new:
```diff
@@ -111,41 +156,152 @@ pipeline {
+
+void cloneRepo(String repo, String dirName=repo) {
+    dir(dirName) {
+        checkout(githubscm.resolveRepository(repo, getGitAuthor(), getBuildBranch(), false))
```
wasn't there a branch problem here ?

Kevin-Mok replied:
That's only for the promote Jenkinsfile since we need to merge a PR. Here, we create a new branch in the "Prepare for PR" stage, so it's not a problem.

Kevin-Mok commented:
Jenkins, retest this.

Kevin-Mok commented on Jenkinsfile.promote:
```diff
@@ -0,0 +1,375 @@
+}
+
+void setDefaultBranch(String repo, String defaultBranch) {
+    withCredentials([usernamePassword(credentialsId: env.AUTHOR_CREDS_ID, usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]) {
```
@radtriste Do we actually need this block?

radtriste replied:
right... not sure we still need it ... GITHUB_TOKEN should be enough
